content/en/blog/2022/tail-sampling/index.md

```diff
@@ -0,0 +1,217 @@
+          status_code: {status_codes: [ERROR]}
+        },
+        {
+          name: randomized-policy,
+          type: probabilistic,
+          probabilistic: {sampling_percentage: 25}
+        },
+      ]
+```
+
+* `tail_sampling` is the name of the processor you’ll use to implement tail
+  sampling.
+* The first three lines are optional configurable settings:
+  * `decision_wait` is the time, in seconds, after the first span of a trace is
+    created before the sampling decision is made. The default is 30 seconds.
+  * `num_traces` is the number of traces to be kept in memory. The default is
+    50,000.
+  * `expected_new_traces_per_sec` is the expected number of new traces, which
+    helps in allocating data structures. The default is 0.
+* `policies` is where you define your sampling policies. There is no default,
+  and this is the only required section of the processor configuration. In this
+  case, two policies are defined:
+  * `status_code`, which is named `errors-policy`, since this example will
+    filter traces with the status code `ERROR`.
+  * `probabilistic`, which is named `randomized-policy`. In addition to
+    filtering all traces with errors, there will also be a randomized sampling
+    of 25% of traces without errors.
+
+The next image is an example of what you might see in your backend if you
+implement this sample configuration.
+
+![Illustration shows sampling based on both traces that have errors as well as randomly selected traces.](backend_visual_errors_and_random.png)
+
+The blue dots and rectangle on the right side indicate that the sampling
+decision occurs at the end of a trace when all the spans for a given request
+have been completed. The green dots represent sampled spans while the gray dots
+represent unsampled spans. Finally, the red dots represent spans where errors
+were detected. With this configuration, you’ll get all traces with errors as
+well as a random sampling of other traces based on the rate we’ve configured. 
+
+If you want to just sample based on a specific filter such as errors, you could
+remove the probabilistic policy.  But having a random sampling of all other
+traces can help surface other issues and give you a broader view of your
+software’s performance and behavior. Here’s what you'll see with only the status
+code policy defined.
+
+![Illustration shows an example of tail sampling based on errors only, and not including random traces.](backend_visual_errors_only.png)
+
+You also have the flexibility to add other policies. Here are a few examples:
+
+* `always_sample`: Sample all traces.
+* `latency`: Sample based on the duration of the trace. For example, you could
+  sample all traces that take longer than 5 seconds.
+* `string_attribute`: Sample based on string attribute values, both exact and
+  regex value matches are supported. For example, you could sample based on
+  specific custom attribute values.
+
+## Potential issues with tail sampling
```

Comment:
I know this wasn't in the NR blog post, but a major thing to consider as well is that otel doesn't propagate metadata that can let a backend re-weight counts (e.g., total events, P95, P99, etc.) - and so you're getting a look at sampled data, but not accurate measurements w.r.t. _all_ data. Sort of a middle ground between sampling as-possible with OTel today and not sampling at all.

Is that something that you'd want to describe in this post as well? I'm fine with it not being here, it's just something that came to mind that isn't mentioned.

Reply:
Interesting point, how much would probabilistic sampling help with that?

Reply:
Unless there's a way to pass a long the effective sample rate, then I can't see how that could help get accurate counts on sampled data, no.

Reply:
Ah, thank you for this! I'm interested in including it, is there anything to expand on further with this point?

Reply:
Yeah, so basically the gist of it is this premise:

> In the monitoring industry there’s a complicated and frustrating conversation that persisted over the years: how do you deal with the enormous volume of data generated by instrumentation? On one side of the aisle, you will find a cohort of vendors and developers telling you that you have to sample data, followed immediately by another group telling you that sampling will ruin the accuracy of incident analysis. They’re both right.

From https://blog.sentry.io/2022/08/23/introducing-dynamic-sampling

If you're only taking 1/5 traces, then unless there's something the backend can use to understand that the data it's operating on is only 1/5 the size of the _real_ data, any measurement (e.g., counting total spans) will be inaccurate. And so if you're working with statistically significant data, then that _might_ be fine for some scenarios, but could be bad for others where you actually do need an accurate accounting of data.

One middle ground here is to have metadata attached to spans/events/etc. that tells a backend what the sample rate is.

Using that example of sampling 1/5 traces, if there's metadata in those sampled traces that says something to the effect of "I'm one out of five!", then a backend can multiple values by 5 to actually be accurate for measures like "total events during this period of time". This is one of the core concepts behind the current sampling SIG that'd getting hammered out, and it's understandably not rolled out in any of the SDK or collector sampling capabilities because of that.

Reply:
@cartermp 

Thank you! I added a paragraph to the last section. Can you let me know if I've summed it up well, or if I should modify it?

>Another limitation to consider is that since OpenTelemetry does not propagate
metadata that would let a backend re-weight counts – for example, P95, P99, and
total events – you’re only getting a look at the measurements of sampled data,
not accurate measurements in regards to _all_ data. Let’s say you’ve configured
your sampler to keep 25% of all traces. If the backend doesn’t know it’s only
operating on 25% of all data, any measurement it produces will be inaccurate.
One way to get around this is to attach metadata to your spans that tells the
backend what the sample rate is, that will allow the backend to then accurately
measure things such as total span count for a given period of time. The Sampling
SIG is currently working on this concept.